### PR TITLE
clp-package: Prevent query scheduler from scheduling two IR extraction jobs targeting the same file split.

### PR DESCRIPTION
**Pull request by haiqi96**
# Description

Currently, if two IR extraction jobs targeting the same file split are submitted simultaneously, both of them will be scheduled by the query scheduler. This behavior could cause a race condition beacuse eventually both jobs will write to the same output file.

This PR adds a dictionary in the query scheduler to track all on-going IR extraction jobs to prevent the same file split being extracted twice. If multiple jobs target the same file split, the scheduler will only submit the first job to the cluster, and let all other jobs with the same target file split id wait for the first job to finish.

This PR also handles another corner case with the following execution order:

1. User submit a job targeting file_split_id 0
2. webui submit a job to the extraction scheduler
3. scheduler submits a job to the clp cluster
4. User submit a job targeting file_split_id 0 again
5. webui submit a job to the extraction scheduler (because 3 hasn't finished yet)
6. clp cluster finishes extraction, writing the entry to mongodb
7. scheduler handles finished job, remove the file_split_id 0 from the active file split list
8. scheduler pulls job submited in 4, resubmits a job to the clp_cluster. (duplication of 3)

To avoid this, before step 8, we let scheduler pull from the mongodb to see if the file split is already available.

Besides the fixes above, the PR also adds indexes to the ir_collection.

# Validation performed
1. Let ir_extraction script submit two jobs simutaneously, confirmed that the second job is marked as running but not submitted to the worker cluster
2. submit multiple jobs targeting the same (orig_file_id, msg_ix) over a short period. confirmed that only the first one was run and the other ones are skipped. Although this is not the exact the same as the corner case being described, they test the same execution flow.

### CODE REVIEW

File: components/clp-py-utils/clp_py_utils/create-results-cache-indexes.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        results_cache_config = ResultsCache.parse_obj(read_yaml_config_file(config_file_path))
+        if results_cache_config is None:
+            raise ValueError(f"results cache configuration file '{parsed_args.config}' is empty.")
+
+        results_cache_uri = results_cache_config.get_uri()
+        ir_collection_name = results_cache_config.ir_collection_name
+
+        with pymongo.MongoClient(results_cache_uri) as results_cache_client:
+            ir_collection = results_cache_client.get_default_database()[ir_collection_name]
+            ir_collection.create_index(["file_split_id"])
```
Comment (haiqi96) [inline components/clp-py-utils/clp_py_utils/create-results-cache-indexes.py:39]: I leave them as constant to follow database creation scripts. If needed, we can create a constant.py for both database and results_cache script

File: components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py [hunk 2]
```diff
@@ -542,10 +547,51 @@ def handle_pending_query_jobs(
...
+                # The second if statement only check if any chunk of the specific file split is extracted, but
+                # it does not guarantee all chunks to be ready because the extraction job could still be running.
+                # However, if there is no running extraction job for the file split, then a single chunk being
+                # available guarantees that all chunks are extracted.
+                if file_split_id in active_extraction_file_splits:
+                    active_extraction_file_splits[file_split_id].append(job_id)
+                    logger.info(
+                        f"Split {file_split_id} is being extracted, mark job {job_id} as running"
+                    )
+                    if not set_job_or_task_status(
```
Comment (haiqi96) [inline components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py:560]: imo, this `if not set_job_or_task_status error` is quite annoying to write and repeated several places in our code. I want to wrap it with a method, but not sure if we want to put in a separate PR.

**Review by junhaoliao — CHANGES_REQUESTED** [review]
The logics match what we discussed offline and seem correct. There're some suggestions about how we can better pass the MongoDB URI to the script.

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py [hunk 3]
```diff
@@ -234,6 +234,61 @@ def create_db_tables(
...
+    component_name = RESULTS_CACHE_COMPONENT_NAME
+    logger.info(f"Creating {component_name} indexes...")
+
+    container_name = f"clp-{component_name}-indexes-creator-{instance_id}"
+
+    # Create results cache config file
+    results_cache_config_filename = f"{container_name}.yml"
+    results_cache_config_file_path = clp_config.logs_directory / results_cache_config_filename
+    with open(results_cache_config_file_path, "w") as f:
+        yaml.safe_dump(container_clp_config.results_cache.dict(), f)
```
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/clp-package-utils/clp_package_utils/scripts/start_clp.py:252]: this is to be consistent with MySQL database flow. but it is true that there's less parameters for results cache. @kirkrodrigues any preference?

File: components/clp-py-utils/clp_py_utils/create-results-cache-indexes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/clp-py-utils/clp_py_utils/create-results-cache-indexes.py:39]: I assume you suggest that we replace this with one create_indexes call with both indexes?

[the remark this answers is not in the document]
Comment (junhaoliao) [reply components/clp-py-utils/clp_py_utils/create-results-cache-indexes.py:39]: Right, like I was thinking we should do something like this?
```
    ir_collection.create_index(["file_split_id", "orig_file_id", "begin_msg_ix", "end_msg_ix"])
```

[the remark this answers is not in the document]
Comment (haiqi96) [reply components/clp-py-utils/clp_py_utils/create-results-cache-indexes.py:39]: If my understanding is correct, this means creating a single index with all the four fields as a tuple.

In our code, we will have two query flow:
the query scheduler will look only for a file_split_id, and the webui will look for something like "orig_file_id = bla and begin_msg_ix < X and X < end_msg_ix". 

so we should create two separate indexes.

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/scripts/start_clp.py:252]: Since we're not using auth for MongoDB, sure, makes sense to simplify things and get rid of the config file. For MySQL, it's different since we don't want to pass credentials through the CLI (we could change that to env-vars, but separate issue).

**Review by junhaoliao — APPROVED** [review]
LGTM. Tested with modified Log Viewer WebUI server which inserted two extract IR jobs quickly one after another, and no job exec failure was observed.

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py [hunk 4]
```diff
@@ -234,6 +234,54 @@ def create_db_tables(
...
+        "--uri", container_clp_config.results_cache.get_uri(),
+        "--ir-collection", container_clp_config.results_cache.ir_collection_name,
+    ]
+    # fmt: on
+
+    cmd = container_start_cmd + create_tables_cmd
+    logger.debug(" ".join(cmd))
+    subprocess.run(cmd, stdout=subprocess.DEVNULL, check=True)
+
+    logger.info(f"Created {component_name} indexes.")
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:282]:
```suggestion
    logger.info(f"Created {component_name} indices.")
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

clp-package: Prevent query scheduler from scheduling two IR extraction jobs targeting the same file split.
